File: microspice_matlab/+parser/parser.m
% Parses the spice netlist
classdef parser < handle
    properties
        file_name       (1, 1)  string = "";         % The input spice file name
        next_line       (1, 1)  int16  = 0;          % The next line number being read
        file_lines      (1, :)  string;              % Each line is separated
        envs            (1, :)  engine.environment;                
        eng             (1, 1)  engine.engine;
    end
    
    methods
        function ret = read(obj, inp_file)
            ret = error.ok;
            if(~isfile(inp_file))
                ret = error.gen_error("File " + inp_file + "does not exist");
                return
            end
            
            obj.file_name   = inp_file;
            file_text       = fileread(obj.file_name);
            obj.next_line   = 1;
            obj.envs        = [engine.environment];
            obj.eng         = engine.engine;
            obj.eng.options("mode") = 1;
            
            obj.eng.set_env(obj.envs(1));
            
            obj.file_lines      = strsplit(file_text,'\n');
        end
        
        function ret = parse(obj)
            max_lines = numel(obj.file_lines);
            
            while(obj.next_line <= max_lines)
                ret = obj.parse_line;
                if(ret.level())
                    return
                end
            end
        end
        
        function ret = parse_line(obj)
            ret = error.ok;
            
            if(numel(obj.file_lines) < obj.next_line)
                ret = error.gen_error("File read completely");
                return
            end
            
            line = split(strip(obj.file_lines(obj.next_line)), '*');
            line = char(line(1));
            
            if(numel(line) == 0)
                obj.next_line = obj.next_line + 1;
                return
            end
            
            switch(lower(line(1)))
                case '*'
                    obj.next_line  = obj.next_line + 1;
                    return
                case '.'
                    ret = obj.parse_command(line);
                    obj.next_line  = obj.next_line + 1;
                    return
                case 'c'
                    e = elements.capacitor;
                    ret = e.read_spice([obj.file_name, obj.next_line, line]);
                case 'r'
                    e = elements.resistor;
                    ret = e.read_spice([obj.file_name, obj.next_line, line]);
                case 'v'
                    
                    pattern = '^(?<id>\w+)\s+(?<n1>\w+)\s+(?<n2>\w+)\s+(?<type>\w+)(\s+)?\(.+\)$';
                    match = regexpi(line, pattern, 'names');
                    
                    if(isempty(match))
                        e = elements.v_const;
                        ret = e.read_spice([obj.file_name, obj.next_line, line]);
                    else
                        switch(lower(match.type))
                            case "pulse"
                                e = elements.v_pulse;
                                ret = e.read_spice([obj.file_name, obj.next_line, line]);
                            case "pwl"
                                e = elements.v_pwl;
                                ret = e.read_spice([obj.file_name, obj.next_line, line]);
                            case "sin"
                                e = elements.v_sin;
                                ret = e.read_spice([obj.file_name, obj.next_line, line]);
                            otherwise
                                ret = error.inp_error.inp_error(obj.file_name, obj.next_line, "Unidentified voltage source type " + match.type);
                        end
                    end
                case 'g'
                    e = elements.vccs;
                    ret = e.read_spice([obj.file_name, obj.next_line, line]);
                otherwise
                    ret = error.inp_error.inp_error(obj.file_name, obj.next_line, "Unindentified prefix");
                    obj.next_line  = obj.next_line + 1;
                    return
            end
            
            if(~ret.level())
                obj.envs(end).add_component(e);
                obj.next_line  = obj.next_line + 1;
            end
        end
        
        function ret = parse_command(obj, line)
            % NOTE : Add error detection here too!
            ret = error.ok;

            cmd_parts = split(line(2:end), ' ');
            cmd_parts(cellfun('isempty', cmd_parts)) = [];
            
            switch(lower(cmd_parts{1}))
                case "option"
                    obj.eng.add_option("option", cmd_parts{2})
                case "print"
                    obj.eng.add_node_print(cmd_parts{2})
                case "tran"
                    obj.eng.add_option("mode", 2)
                    step_size = parser.parse_number(cmd_parts{2});
                    end_time  = parser.parse_number(cmd_parts{3});
                    
                    obj.eng.add_option("step_size", step_size);
                    obj.eng.add_option("end_time", end_time);
                case "alter"
                    obj.envs(end+1) = copy(obj.envs(end));
            end
        end
    end
end

File: microspice/parser.py
import microspice.engine        as engine
import microspice.error.error   as error
import microspice.environment   as environment
import microspice.elements      as elements
from   microspice.utils         import *

import os.path
import re
import copy

# This class parses a file and populates the engine and environment
#       with components, connectivity and simulation settings
class Parser:
    def __init__(self):
        self.file_name  = ""
        self.next_line  = 0
        self.file_lines = []
        self.envs       = []
        self.eng        = engine.Engine()

    # Read a netlist into buffer and initialize variables
    def read(self, inp_file):
        ret = error.OkError()

        # Check if the file exists
        if not os.path.isfile(inp_file):
            ret = error.GenError("File " + inp_file + " does not exist")
            return ret

        # Set the filename and line counts (used for debugging)
        self.file_name = inp_file
        self.next_line = 1
        
        # Initialize the engine and environment 
        self.envs      = [environment.Environment()]        
        self.eng       = engine.Engine()
        self.eng.options["mode"] = 1
        self.eng.set_env(self.envs[0])

        # Read the file
        with open(self.file_name, 'r') as file:
            self.file_lines = file.read().split('\n')
        
        return ret

    # Parse the whole netlist
    def parse(self):
        max_lines = len(self.file_lines)

        # Iterate through all the lines
        while self.next_line <= max_lines:
            ret = self.parse_line()
            if ret.level():
                return ret
            
        return ret

    # Parse a single line of the spice netlist
    def parse_line(self):
        ret = error.OkError()

        # Last line (should not happen unless function is called after running "parse()")
        if len(self.file_lines) < self.next_line:
            ret = error.GenError("File read completely")
            return ret

        # Read file, remove comments and leading and trailing whitespace
        line = self.file_lines[self.next_line - 1].split('*')[0].strip()
        if len(line) == 0:
            self.next_line += 1
            return ret

        # Decide the element type or the command 
        switch_case = line[0].lower()
        # Comment
        if switch_case == '*':
            self.next_line += 1
            return ret
        # Command
        elif switch_case == '.':
            ret = self.parse_command(line)
            self.next_line += 1
            return ret
        # Capacitor
        elif switch_case == 'c':
            e = elements.Capacitor()
            ret = e.read_spice([self.file_name, self.next_line, line])
        # Resistor
        elif switch_case == 'r':
            e = elements.Resistor()
            ret = e.read_spice([self.file_name, self.next_line, line])
        # Voltage source
        elif switch_case == 'v':
            pattern = r'^(?P<id>\w+)\s+(?P<n1>\w+)\s+(?P<n2>\w+)\s+(?P<type>\w+)(\s+)?\(.+\)$'
            match = re.match(pattern, line)

            if match is None:
                e = elements.VConst()
                ret = e.read_spice([self.file_name, self.next_line, line])
            else:
                voltage_type = match.group('type').lower()
                if voltage_type == "pulse":
                    e = elements.VPulse()
                    ret = e.read_spice([self.file_name, self.next_line, line])
                elif voltage_type == "pwl":
                    e = elements.VPWL()
                    ret = e.read_spice([self.file_name, self.next_line, line])
                elif voltage_type == "sin":
                    e = elements.VSin()
                    ret = e.read_spice([self.file_name, self.next_line, line])
                else:
                    ret = error.InpError(self.file_name, self.next_line,
                                         "Unidentified voltage source type " + voltage_type)
        # VCCS (Voltage controlled current source)
        elif switch_case == 'g':
            e = elements.VCCS()
            ret = e.read_spice([self.file_name, self.next_line, line])
        # Not implemented!
        else:
            ret = error.InpError(self.file_name, self.next_line, "Unidentified prefix")
            self.next_line += 1
            return ret

        # Error detection
        if not ret.level():
            self.envs[-1].add_component(e)
            self.next_line += 1

        return ret

    # Parse a spice command (starts with .)
    def parse_command(self, line):
        ret = error.OkError()

        # Extract the components of the command (separated by spaces)
        cmd_parts = line[1:].split(' ')
        cmd_parts = [part for part in cmd_parts if part != ""]

        switch_case = cmd_parts[0].lower()

        # SPICE commands
        if switch_case == "option":
            # Options are added to the options dictionary in the engine
            self.eng.add_option("option", cmd_parts[1])
        
        elif switch_case == "print":
            # Nodes to be printed are added to the print list in the engine
            self.eng.add_node_print(cmd_parts[1])
        
        elif switch_case == "tran":
            # For setting transient mode, the "mode" option is set to 2 and other
            #       settings are added to the options dictionary
            step_size   = parse_number(cmd_parts[1])
            end_time    = parse_number(cmd_parts[2])
            
            self.eng.add_option("mode", 2)
            self.eng.add_option("step_size", step_size)
            self.eng.add_option("end_time", end_time)

        elif switch_case == "alter":
            # The alter command sets up one more simulation by creating a copy of the most recent environment
            self.envs.append(copy.deepcopy(self.envs[-1]))

        return ret
